Adds power tools and the framework for adding faster tools that can swap heads.

l3lb0t opened the pull request:
:cl: Karl Johansson
rscadd: Adds the hand drill, a portable hand drill that can swap between a bolt and screw head. Does (most) things twice as fast! Sprites done by Azlan.
rscadd: Adds the hydraulic prying tool, a tool that can swap between prying and cutting heads. Does (most) things twice as fast! Sprites also done by Azlan.
tweak: Redoes tool logic to allow for tools that perform actions faster or slower than the baseline.
maptweak: Changes the toolbelt in the Chief Engineer's locker to include a hand drill and hydraulic prying tool.
/:cl:

### Frequently Asked Questions

**What?**
This pull request adds a set of high-quality tools that are given to the Chief Engineer and are also printable by research. It also sets up a system for high or low quality tools that can do tasks at different speeds to the baseline.

**Why?** 
Personally, I thought it was weird how power tools had yet to find their way into the hands of the SEV Torch's engineers already, considering that the year is supposed to be 2311. Also, it makes adding these kinds of items way easier for anyone else in the future.

**Who's to blame for this?**
Karl Johansson - That's me, I did most of the coding. Yes, I know it's bad.
Azlan/SomeAngryMiner - Did the spritework and did a lot of other stuff behind the scenes, like helping me rebase my 22 unhinged commits into 6 sensible ones.

**Did you even test this?**
I've tested it about as much as I feasibly can. They spawn properly, work properly, and seem balanced. I think if there was a significant problem with any of the interactions, they would be caught as an error in compiling, anyways.

l3lb0t commented:
I had _really_ hoped that error was to do with VSCode being weird. It runs fine when I test it.

l3lb0t commented:
Not entirely certain what I'm supposed to do about using .len one too many times. I'm pretty sure I use it exactly once.

MuckerMayhem commented on code/game/objects/items/weapons/tools/swapper.dm:
```diff
@@ -0,0 +1,105 @@
+	var/tool_index = tool_list.Find(active_tool)
+	previous_tool = tool_index
+	tool_index = tool_index + 1
+	if (tool_index > tool_list.len)
```
```suggestion
	if (tool_index > length(tool_list))
```

MuckerMayhem commented on code/game/objects/items/weapons/tools/swapper.dm:
```diff
@@ -0,0 +1,105 @@
+	if (tool_index > length(tool_list))
+		tool_index = 1
+	active_tool = tool_list[tool_index]
+	to_chat(user, "<span class='notice'>You attach the [active_tool] to [src].</span>")
```
```suggestion
	to_chat(user, SPAN_NOTICE("You attach the [active_tool] to [src]."))
```

MuckerMayhem commented on code/game/objects/items/weapons/tools/swapper.dm:
```diff
@@ -0,0 +1,105 @@
+		sharp = FALSE
+	else
+		sharp = TRUE
+	if (active_tool == CUTTER_HEAD) //but also so you can behead people with the cutters because why not
```
```suggestion
	if (active_tool == CUTTER_HEAD)

MuckerMayhem commented on code/game/objects/items/weapons/tools/swapper.dm:
```diff
@@ -0,0 +1,105 @@
+#define PULSE_HEAD "electrode head"
+#define SCREW_HEAD "screw head"
+
+/obj/item/swapper
+	name = "the concept of a tool that can swap between multiple heads"
+	desc = "This is a master item, berate the admin or mapper that spawned this!"
```
```suggestion
/obj/item/swapper
 	abstract_type = /obj/item/swapper     
	name = "the concept of a tool that can swap between multiple heads"
	desc = "This is a master item, berate the admin or mapper that spawned this!"
```
Defining this as an abstract will prevent it from spawning, and is good practice.

l3lb0t replied:
This seems to cause a lot of errors, so I presume more needs to be done here, but I don't know what.

l3lb0t replied:
Tried this again, I think it was just an indentation issue.

l3lb0t replied:
https://github.com/Baystation12/Baystation12/pull/33539/commits/34d89244db3ceb235a265363e029c8dfa40b8396

l3lb0t commented:
That seems to have broken things.

l3lb0t commented:
Undefining it as an abstract seems to have fixed that, which sucks. There's probably a better fix, but it's also 0:48 for me, so I'll come back to this tomorrow.

SierraKomodo commented on code/_helpers/unsorted.dm:
```diff
@@ -886,6 +886,8 @@ GLOBAL_LIST_INIT(duplicate_object_disallowed_vars, list(
 /obj/item/crowbar/istool()
 	return TRUE
 
+/obj/item/swapper/istool()
+	return TRUE
```
Move this to the swapper.dm file, where everything else for that class is being defined.

SierraKomodo commented on code/game/objects/items.dm:
```diff
@@ -22,6 +22,7 @@
 	var/force = 0
 	var/attack_cooldown = DEFAULT_WEAPON_COOLDOWN
 	var/melee_accuracy_bonus = 0
+	var/toolspeed = 1 //Used to determine how fast an item completes a certain action. Lower values are faster.
```
Use DMDoc formatting when documenting properties.

See https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks for specifics

```suggestion
	/// Used to determine how fast an item completes a certain action. Lower values are faster.
	var/toolspeed = 1
```

SierraKomodo commented on code/modules/tables/tables.dm:
```diff
@@ -306,27 +306,28 @@
-	playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
-	if(!do_after(user, 2 SECONDS, src, DO_REPAIR_CONSTRUCT))
-		manipulating = 0
+/obj/structure/table/proc/remove_reinforced(obj/item/S, mob/user)
+	reinforced = common_material_remove(user, reinforced, S, "reinforcements", "screws", 'sound/items/Screwdriver.ogg')
+
+/obj/structure/table/proc/remove_material(obj/item/W, mob/user)
+	material = common_material_remove(user, material, W, "plating", "bolts", 'sound/items/Ratchet.ogg')
+
```
I would prefer keeping the shorter/longer time between material and reinforced, and applying the toolspeed to the parameters of `common_material_remove()` instead.

```suggestion
/obj/structure/table/proc/remove_reinforced(obj/item/S, mob/user)
	reinforced = common_material_remove(user, reinforced, S.toolspeed * 40, "reinforcements", "screws", 'sound/items/Screwdriver.ogg')

/obj/structure/table/proc/remove_material(obj/item/W, mob/user)
	material = common_material_remove(user, material, W.toolspeed * 20, "plating", "bolts", 'sound/items/Ratchet.ogg')

```

l3lb0t replied:
I would, too, but in testing this appeared to just not work. I'm not certain why. I'll test this out locally again.

l3lb0t replied:
Tested this again. It causes materials to be removed instantly from tables when you use a wrench of any kind on them.

l3lb0t replied:
I reviewed the code. Looks like the third variable in common_material_remove is completely unused except for me modifying it to give the toolspeed of the item being used. I'm genuinely not sure if there ever was a time difference between reinforced and material dismantling.
I'll resolve this for now, but I'll reopen it if you need me to.

l3lb0t replied:
I reviewed the code. Looks like the third variable in common_material_remove is completely unused except for me modifying it to give the toolspeed of the item being used. I'm genuinely not sure if there ever was a time difference between reinforced and material dismantling.
However, I'm also not going to argue with the headdev on how their codebase works, so I partially rewrote tablecode to make this work. I'll resolve this for now, unless you want it reopened.
https://github.com/Baystation12/Baystation12/pull/33539/commits/b39ffc2983eb5a39375f7953b7c99185e7143f94
Seems to work fine in local testing.

SierraKomodo commented on code/modules/tables/tables.dm:
```diff
@@ -306,27 +306,28 @@
+		reset_mobs_offset()
+		if(manipulating) return
+		manipulating = 1
+		user.visible_message(SPAN_NOTICE("\The [user] begins dismantling \the [src]."),
+									SPAN_NOTICE("You begin dismantling \the [src]."))
+		playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
+		if(!do_after(user, (W.toolspeed * 2) SECONDS, src, DO_REPAIR_CONSTRUCT))
+			manipulating = 0
+			return
+		user.visible_message(SPAN_NOTICE("\The [user] dismantles \the [src]."),
+									SPAN_NOTICE("You dismantle \the [src]."))
+		new /obj/item/stack/material/steel(src.loc)
+		qdel(src)
 		return
-	user.visible_message(SPAN_NOTICE("\The [user] dismantles \the [src]."),
-	                              SPAN_NOTICE("You dismantle \the [src]."))
-	new /obj/item/stack/material/steel(src.loc)
-	qdel(src)
-	return
 
```
Return early instead of nesting.

```suggestion
/obj/structure/table/proc/dismantle(obj/item/W, mob/user)
	if (!isWrench(W))
		return
    
	reset_mobs_offset()
	if(manipulating) return
	manipulating = 1
	user.visible_message(SPAN_NOTICE("\The [user] begins dismantling \the [src]."),
								SPAN_NOTICE("You begin dismantling \the [src]."))
	playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
	if(!do_after(user, (W.toolspeed * 2) SECONDS, src, DO_REPAIR_CONSTRUCT))
		manipulating = 0
		return
	user.visible_message(SPAN_NOTICE("\The [user] dismantles \the [src]."),
								SPAN_NOTICE("You dismantle \the [src]."))
	new /obj/item/stack/material/steel(src.loc)
	qdel(src)
	return

```

l3lb0t replied:
> I don't think I wrote most of this, but I suppose I might as well optimize while I'm here.

Nope, I did actually do this. Optimization time.

l3lb0t commented:
Everything has been addressed, I think.

MuckerMayhem reviewed (changes requested):
needs squashing

l3lb0t commented:
nice and squashed

l3lb0t commented:
And the static type issue's back. Weird.

MuckerMayhem commented:
Needs more squashing. Only give the minimum amount of commits you need, and make sure they have relevant names.

l3lb0t committed:
Adds a variable for how fast tools do things

Called 'toolspeed'. It multiplicatively defines how long it takes to complete a certain tool-based action. Lower values are faster.

l3lb0t commented:
Whoops. Anyways, while I accidentally closed the PR, I also squashed it down to 2 commits. Theoretically it could go down to one but I don't think it would be helpful to group the items I'm adding in with all the random other interactions that aren't necessarily related.

l3lb0t commented:
Did a final run through. There's a bug, so I'm going to ask for this not to be merged until I address that.

l3lb0t committed:
Adds power tools

Adds two tools that can swap between different heads for a range of uses. They work faster than normal tools. They can be made by R&D but are also given to the CE at roundstart.

l3lb0t commented:
Okay, looks like AwaySites somehow broke. No idea how that happened.
